browser/page_mapping.go

```diff
@@ -423,70 +423,87 @@ func mapPage(vu moduleVU, p *common.Page) mapping { //nolint:gocognit,cyclop
 func mapPageOn(vu moduleVU, p *common.Page) func(common.PageOnEventName, sobek.Callable) error { //nolint:funlen
 	rt := vu.Runtime()
 
+	pageOnEvents := map[common.PageOnEventName]struct {
+		mapp func(vu moduleVU, event common.PageOnEvent) mapping
+		prep func() error
+		wait bool // should we wait for the handler to complete?
+	}{
+		common.EventPageConsoleAPICalled: {
+			mapp: mapConsoleMessage,
+			wait: false,
+		},
+		common.EventPageMetricCalled: {
+			mapp: mapMetricEvent,
+			prep: prepK6BrowserRegExChecker(rt),
+			wait: true,
+		},
+	}
+
 	return func(eventName common.PageOnEventName, handleEvent sobek.Callable) error {
-		tq := vu.taskQueueRegistry.get(vu.Context(), p.TargetID())
+		pageOnEvent, ok := pageOnEvents[eventName]
+		if !ok {
+			return fmt.Errorf("unknown page on event: %q", eventName)
+		}
 
-		onEventPageConsoleAPICalled := func(event common.PageOnEvent) {
-			tq.Queue(func() error {
-				mapping := mapConsoleMessage(vu, event)
-				_, err := handleEvent(sobek.Undefined(), rt.ToValue(mapping))
-				if err != nil {
-					return fmt.Errorf("executing page.on handler: %w", err)
-				}
-				return nil
-			})
+		// Prepare the environment for the page.on event handler if necessary.
+		if pageOnEvent.prep != nil {
+			if err := pageOnEvent.prep(); err != nil {
+				return fmt.Errorf("initiating page.on('%s'): %w", eventName, err)
+			}
 		}
 
-		onEventPageMetricCalled := func(event common.PageOnEvent) {
-			// The function on the taskqueue runs in its own goroutine
-			// so we need to use a channel to wait for it to complete
-			// since we're waiting for updates from the handler which
-			// will be written to the ExportedMetric.
+		// Queue the event handler in the task queue.
+		// Wait for the handler to complete if necessary.
+		tq := vu.taskQueueRegistry.get(vu.Context(), p.TargetID())
+		queueHandler := func(event common.PageOnEvent) {
 			done := make(chan struct{})
+
 			tq.Queue(func() error {
 				defer close(done)
 
-				mapping := mapMetricEvent(vu, event)
-				_, err := handleEvent(sobek.Undefined(), rt.ToValue(mapping))
+				mapping := pageOnEvent.mapp(vu, event)
```

Comment:
I know this isn't due to your change, but I'm wondering if we can perform the mapping outside the taskqueue? 🤔

Reply:
Nice idea 👍